Best Practices : Versioning GraphQL APIs (#884)
graphql/graphql.github.io: site/learn/BestPractice-Versioning.md

### Versioning of REST
Now let’s have a look on versioning of RESTful API. You’ll wonder why we are talking about RESTful APIs. But gradually you will understand how versioning of  GraphQL is much simpler and beautiful than REST. There are many best practices and strategies for REST API versioning such as

* URI Versioning : /api/v1
* Query Parameters :  /api?version=1
* Custom Headers : Accept-version : v1

In most of the cases, uri versioning is used for versioning REST API. The client uses a specific version of API for implementations. 

### Let's have a look at GraphQL
For GraphQL, you introduce or deprecate the fields or schema instead of versioning. While using GraphQL, one key point is that the query itself decides what response client wants. So there will be only one endpoint while using GraphQL API. We can say that instead of versioning, we are just modifying the existing schema. 

Let’s understand by an example. 
We have a schema named user, which had fields such as id, firstName and lastName. 
```graphql
type User {
	id: Int!,
	firstName : String!,

> 	lastName : String!

Author: I'll reformat the file through prettier in VS code. I hope that will fix all format issues.

}
```

In the above schema all fields are required. But in the next version we can add a variable and can also change the requirements. Such as :
```graphql
type User {
	Id : Int!,
	firstName : String!,
	middleName : String!

> 	lastName : String,

Author: Agreed. It was just a typo by me. Fixed in next commit.

}
```

> Here we have added the middleName parameter which is mandatory and have changed lastName to optional.

Reviewer: You can change an optional field to required, but you cannot change a required field to optional (that's a breaking change). The term `field` is the correct term to use here (rather than `parameter`).

```suggestion
Here we have added the `middleName` field which is optional.
```

* The best practice is to use “@deprecated(reason)” defined in the schema. By using this the client can know which filed to use and which one is deprecated. This will not define the version number as the semantic versioning does. The deprecated fields will be shown by the GraphQL Explorer.

### More info on Versioning
There are few other best practices for versioning GraphQL APIs.
Version constraints can be added to the query which will tell that the field will be resolved by which version. Using version constraints we can access different fields of different versions. For example, the query will look like : 
```graphql
query {
	account(id : 5){
		old : profilePhoto(versionContraint : “^0.1”)
		new : profilePhoto(versionContraint : “^0.2”)
	}
}
```

> We can define the versions in the schema and query can be written as above to access the fields of different versions.

Author: It's actually tagging+semantic versioning. I do not remember where I read this, but I'll try to find it.

> We can define the versions in the schema and query can be written as above to access the fields of different versions.

Author: @benjie Should I keep this point?

```graphql
type  User{
	id: String!
	name: String!
	Username : String!
	birthDate : DateTime!
	lastLoginHistory : [DateTime]!
}

```
In the above schema we have one field named lastLoginHistory which stores the DateTime fields as a list. Now there needs some changes in the schema and some additional data needs to be added in lastLoginHistory. For that we need to add another field to the User schema. These changes will add many fields and this could make the schema lengthier and worse.

But rather than this, we can build a schema for storing lastLoginHistory data and we can use that schema referenced to the User schema.
```graphql
type NewDateTime : {
	lastLoginTime : DateTime!
	lastLoginPeriodLength : Time!
}
```
``` graphql
type User{
	id: String!
	name: String!
	Username : String!
	birthDate : DateTime!

> 	lastLoginHistory : NewDateTime!

Reviewer 2: but... this would be a breaking change for clients?

Author: How? I think we can evolve schema by referencing schemas to others. Any specifics why this will be a breaking change?
